broker/fragment/store_azure.go

```diff
@@ -324,7 +403,16 @@ func (a *azureBackend) buildBlobURL(cfg AzureStoreConfig, client pipeline.Pipeli
 }
 
 // Cache UserDelegationCredentials and refresh them when needed
-func (a *azureBackend) getUserDelegationCredential() (*service.UserDelegationCredential, error) {
+func (a *azureBackend) getUserDelegationCredential(storageMapping *url.URL) (*service.UserDelegationCredential, error) {
+	if a.udcs == nil {
+		a.udcs = make(map[string]UdcAndExp)
+	}
+	var cfg, err = parseAzureStorageMapping(storageMapping)
+	if err != nil {
+		return nil, err
+	}
+	var udc, hasCachedUdc = a.udcs[cfg.accountTenantID]
```

Comment:
This, and the initialization above, need to be guarded by a mutex.

Please carefully review everywhere that you're mutating and accessing these maps. It always always needs to be guarded.

I would recommend wrapping it in a brief critical section here which handles initialization and pulls out `udc` & `hasCachedUdc`. Then, below where you upsert, lock there as well.

I would rename `clientMu` to just `mu`, and essentially any time you're touching state attached to `a`, it needs to be within the scope of a briefly-held lock. It's perfectly fine, i'd think, if multiple goroutines each independently fetch credentials and overwrite each other :shrug: so you don't need to hold locks for long.

Reply:
Okay, we talked on a VC and my understanding here is: any area that reads _or mutates_ state, and can be entered from multiple threads/goroutines needs a critical section. I believe I got all of them here.